run/aix2john.py

```diff
@@ -1,87 +1,172 @@
 #!/usr/bin/env python
 
-# This script extracts hashes from AIX /etc/security/passwd
-# cat /etc/security/passwd
-# root:
-#         password = mrXXXXXXXXXX
-#         lastupdate = 1343960660
-#         flags =
-#
-# admin:
-#         password = oiXXXXXXXXXX
-#         lastupdate = 1339748349
-#         flags =
-# ...
-
-# Note: Some of this functionality is also covered by the unshadow program.
-
-
+"""
+This Python script is designed to parse and extract user credentials data from
+an IBM AIX /etc/security/passwd file. The credentials consist of the username
+and the corresponding password hashes.
+
+The parser handles different password hash types including DES, MD5, SMD5,
+SHA256, and SHA512, with automatic recognition based on the hash length. In
+addition to the normal operation, the script provides an option to list all
+users, even those with locked or unset passwords.
+
+Features:
+- Automatic password hash type identification.
+- Option to display all users, regardless of account status.
+- Error handling for file opening exceptions.
+
+Usage:
+The script can be used from the command line with the following options:
+
+    python aix2john.py -f [filepath] : To specify a file path for processing.
+    The default is /etc/security/passwd.
+    
+    python aix2john.py -a : To include all users, including those with locked
+    or unset passwords.
+
+The script prints the username, the password hash, and the password hash type
+for each user in a formatted way.
+
+Example /etc/security/passwd file:
+
+        guest:
+        password = *
+
+        paul:
+        password = YFf0/OmVZz6tQ
+        lastupdate = 1026394230
+        flags =
+
+        jonas:
+        password = 5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8
+        lastupdate = 1026394230
+        flags =
+
+Compatibility:
+The script is compatible with both Python 2 and Python 3 environments.
+
+Security and Compliance:
+This script handles sensitive data and should be used with appropriate caution.
+Ensure your usage of this script complies with all relevant laws and
+regulations, and refrain from using it for unauthorized access to password
+data.
+"""
+
+from __future__ import print_function
 import argparse
-import binascii
 import sys
+import logging
+import io
 import re
 
+# Configure logging
+logging.basicConfig(level=logging.ERROR, format="%(levelname)s: %(message)s")
+
+# Define password types and associated lengths
+PASSWORD_TYPES = {
+    "des": 13,
+    "md5": 32,
+    "smd5": 46,
```

Comment:
Why 46? The script previously checked for 37, and the smd5 samples we have are 37.